### PR TITLE
Ginkgo: Migrated headless service tests

### LINKED ISSUE #1985: Add ginkgo version of tests added in #1965
Tests are verifying whether toServices egress rules are translated into toCIDR egress rules.

@ianvernon

### PR DESCRIPTION
**Pull request by eloycoto**
Migrated `04-toservices-test.sh` to Ginkgo test framework

Fix #1985

PS: I know that all these manifest files duplicated in two folders are not the best way. I have a small snippet for fixing this in a better way, but I need to test a bit deeper, so that will be fixed in the coming days. 

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>

### CODE REVIEW

File: test/helpers/kubectl.go [hunk 1]
```diff
@@ -105,6 +105,19 @@ func (kub *Kubectl) GetPods(namespace string, filter string) *CmdRes {
 	return kub.Exec(fmt.Sprintf("%s -n %s get pods %s -o json", KubectlCmd, namespace, filter))
 }
 
+// GetPodsNodes returns a map with pod name as a key and node name as value. It
+// get only pods in the given namespace that match the provided filter.
```
Comment (ianvernon) [inline test/helpers/kubectl.go]: It only gets

File: test/k8sT/Services.go [hunk 2]
```diff
@@ -146,4 +147,105 @@ var _ = Describe("K8sServicesTest", func() {
...
+				data, err := kubectl.GetPodNames(helpers.DefaultNamespace, "zgroup=headless")
+				if err != nil {
+					return false
+				}
+				if len(data) == 0 {
+					return true
+				}
+				return false
+			}
+			err := helpers.WithTimeout(waitFinish, "cannot finish containers",
```
Comment (ianvernon) [inline test/k8sT/Services.go]: change to "cannot finish deleting containers"

**Review by ianvernon — APPROVED** [review]
Minor nitpicks, doesn't need another round of review from me. @nebril please confirm that this test adequately migrates the test from the bash scripts, as you are the author of the original test. Thanks!

File: test/k8sT/Services.go [hunk 3]
```diff
@@ -146,4 +147,105 @@ var _ = Describe("K8sServicesTest", func() {
...
+
+			res = kubectl.Apply(policyPath)
+			res.ExpectSuccess()
+
+			validateEgress()
+
+		})
+
+		It("To Services first policy", func() {
+			res := kubectl.Apply(podPath)
```
Comment (nebril) [inline test/k8sT/Services.go]: Please move pod creation to BeforeEach and ensure it is created to be consistent with original test behaviour.

File: test/k8sT/Services.go [hunk 4]
```diff
@@ -146,4 +147,105 @@ var _ = Describe("K8sServicesTest", func() {
...
+			data, err := res.Filter(`{[0].policy.cidr-policy.egress}`)
+			Expect(err).To(BeNil())
+			Expect(data.String()).To(ContainSubstring(expectedCIDR))
+		}
+
+		It("To Services first endpoint creation", func() {
+			res := kubectl.Apply(endpointPath)
+			res.ExpectSuccess()
+
+			res = kubectl.Apply(podPath)
```
Comment (nebril) [inline test/k8sT/Services.go]: Please move pod creation to BeforeEach and ensure it is created to be consistent with original test behaviour.

**Review by nebril — CHANGES_REQUESTED** [review]
Looks good overall!

Just need to move pod creation to BeforeEach (it's not part of the test, it's part of test scenario).

File: test/k8sT/Services.go (hunk 4 above)
Comment (eloycoto) [reply test/k8sT/Services.go]: Hi @nebril 

Quick one, I made like this that is how the original test is made. First endpoint -> Pods and after that the other way around.

https://github.com/cilium/cilium/blob/master/tests/k8s/tests/04-toservices-test.sh#L66
https://github.com/cilium/cilium/blob/master/tests/k8s/tests/04-toservices-test.sh#L97

**Comment by eloycoto** [comment]
Hi @nebril any comment about my comments? 

Regards.

**Comment by eloycoto** [comment]
@nebril  comments fixed + I rebased the branch. 

Regards

**Review by nebril — APPROVED** [review]

**Pull request merged**